### PR TITLE
[video_player] Fix playback speed resetting on iOS

### PR DESCRIPTION
**Pull request by alexagat**
This stabilizes the behavior of the `video_player` package such that calling `seekTo` does not reset the playback speed to `1.0` on the iOS platform.

*List which issues are fixed by this PR. You must list at least one issue.*
* https://github.com/flutter/flutter/issues/71264

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by alexagat** [comment]
@googlebot I fixed it

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m [hunk 1]
```diff
@@ -242,6 +244,7 @@ - (instancetype)initWithPlayerItem:(AVPlayerItem*)item frameUpdater:(FLTFrameUpd
 
   _player = [AVPlayer playerWithPlayerItem:item];
   _player.actionAtItemEnd = AVPlayerActionAtItemEndNone;
+  [self setInternalPlaybackSpeed:1.0f];
```
Comment (stuartmorgan) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:247]: Why is this needed on init?

Also, per style guide, this should not call self: https://google.github.io/styleguide/objcguide.html#avoid-messaging-the-current-object-within-initializers-and--dealloc
(There are existing violations, but they are wrong, and we shouldn't add more.)

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m [hunk 2]
```diff
@@ -310,6 +313,9 @@ - (void)updatePlayingState {
   }
   if (_isPlaying) {
     [_player play];
+
+    // fix to always set playback speed accurately
+    [self setPlaybackSpeed:_internalPlaybackSpeed];
```
Comment (stuartmorgan) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:318]: Is this always needed here? My impression from comments in the issue was that calling it from the completion handler of the seek method would likely be enough.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m [hunk 3]
```diff
@@ -369,6 +375,10 @@ - (void)setIsLooping:(bool)isLooping {
   _isLooping = isLooping;
 }
 
+- (void)setInternalPlaybackSpeed:(double)speed {
```
Comment (stuartmorgan) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:378]: Why are you overriding a property setter with the default implementation of a property setter.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Comments inline, but this will also need tests (and a version update), per the checklist and contribution guide.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m (hunk 1 above)
Comment (alexagat) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:247]: Good point, this is not needed and I will remove.

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m (hunk 3 above)
Comment (alexagat) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:378]: I was copying the pattern I saw above, but after referring to [Apple docs](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjectiveC/Chapters/ocProperties.html), I see that property declarations give us the getter/setters for free. Nice! Will remove

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m (hunk 2 above)
Comment (alexagat) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:316]: @stuartmorgan which completion handler are you referring to? I only see a `seekTo` method in the file.

Comment (stuartmorgan) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:316]: See https://github.com/flutter/flutter/issues/71264#issuecomment-734656335

Comment (alexagat) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:316]: @stuartmorgan I see. I tried moving the call inside the completionHandler block of the `seekTo` method as follows:

```
[_player seekToTime:CMTimeMake(location, 1000)
      toleranceBefore:kCMTimeZero
       toleranceAfter:kCMTimeZero
    completionHandler:^(BOOL finished) {
    [self setPlaybackSpeed:_internalPlaybackSpeed];
  }];
```

but the issue of the playback speed resetting to `1.0` came back. Therefore that is not equivalent.

Comment (stuartmorgan) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:316]: Interesting. So presumably then it's the `playbackLikelyToKeepUpContext` and/or `AVPlayerItemStatusReadyToPlay` codepath that's resolving this? In which case perhaps the underlying issue is that AVPlayer automatically resets the playback speed when the buffer is insufficient.

**Comment by alexagat** [comment]
@stuartmorgan  / @creativecreatorormaybenot I've realized that while this modification fixes the playback speed issue it causes a negative side-effect. When an unbuffered range of the video stream is played back, the video player can lose volume (go mute) and become unstable. Once it gets in this state it does not recover. Therefore looks like there's still more work, unfortunately.

**Comment by creativecreatorormaybenot** [comment]
@alexagat if you take a look at my PR (#3404), it _does_ fix the issue. Can we not just perform the same action on the native side?

Because at the end of the day, I am only making a method channel call..

**Comment by stuartmorgan** [comment]
> When an unbuffered range of the video stream is played back, the video player can lose volume (go mute) and become unstable.

Does setting it back only for `playbackLikelyToKeepUpContext` work better, or is that already the only place it was being called in practice?

> @alexagat if you take a look at my PR (#3404), it _does_ fix the issue.

It is very likely that your approach would racily have the same behavior. There's no guarantee that some fraction of 500ms is enough time to create sufficient buffer to avoid an overrun.

**Comment by alexagat** [comment]
@creativecreatorormaybenot you are correct your PR does fix this issue on the iOS platform and does not add any of the negative side-effects the native fix above introduced. I played around for a bit but didn't come up with a native-only solution here. For now I've added your code to my fork so we can patch this issue for users.

**Comment by creativecreatorormaybenot** [comment]
@alexagat yeah, that is what we are also doing in production, using our own fork of the plugin.

**Comment by stuartmorgan** [comment]
If there's no way to automatically handle this in the plugin that doesn't have the potential to cause playback issues, the other option would be to add an event that would report playback speed changes back to the Dart side, so that UIs could update, and users could change it back themselves.

But I would expect that there's a way to determine the relevant state to handle this correctly. Does observing `timeControlStatus` give any indication as to what's happening?

**Comment by creativecreatorormaybenot** [comment]
@stuartmorgan I have one concern with this part:

> If there's no way to automatically handle this in the plugin that doesn't have the potential to cause playback issues, the other option would be to add an event that would report playback speed changes back to the Dart side, so that UIs could update, and users could change it back themselves.

You were saying that my fix was a hack. But your proposal is not? If I look at this from a different perspective, this means that users of the problem would need to integrate the hack (changing the playback speed after the change has been reported to the Dart side).  
I see how this would be better in some odd way, but then again I really feel like this is even worse than just using the not-so-great fix because it also makes for a very bad API IMHO.

> But I would expect that there's a way to determine the relevant state to handle this correctly.

That would be my preferred choice as well..

**Comment by stuartmorgan** [comment]
> You were saying that my fix was a hack. But your proposal is not?

There are two possibilities here:
1. There is a reliable programatic way to determine when it is safe to automatically set the playback rate back to the desired value. If there is, we should do that.
2. There is no reliable way to do it automatically, such that any attempt to do so may break playback in the way described above. In this case I don't see any alternative to simply communicating the state back to the UI so that users can control it.

Your PR was sort of attempting to do 1, but in a very hacky way. I'm saying that either we need to do 1 correctly, or if 1 is impossible we should do 2. If 2 is the only option it's not a hack, it a reflection of the limitations of the platform that the plugin is wrapping.

**Comment by alexagat** [comment]
@stuartmorgan I am not planning to update this PR and it should probably be closed. I never got the iOS implementation to work as desired and ended up using the code change from https://github.com/flutter/plugins/pull/3404 to resolve the bug.

**Comment by creativecreatorormaybenot** [comment]
Which is why is still think it should be merged 👍 It is not *perfect* but the best from a cost <> value perspective.

**Comment by stuartmorgan** [comment]
Closing per the above; thanks for the investigation so far. This should serve as a good starting point for future attempts to fix the issue.
